src/passes/GlobalTypeOptimization.cpp

```diff
@@ -484,10 +487,22 @@ struct GlobalTypeOptimization : public Pass {
   void updateTypes(Module& wasm) {
     class TypeRewriter : public GlobalTypeRewriter {
       GlobalTypeOptimization& parent;
+      InsertOrderedMap<HeapType, Index>::iterator placeholderIt;
 
     public:
       TypeRewriter(Module& wasm, GlobalTypeOptimization& parent)
-        : GlobalTypeRewriter(wasm), parent(parent) {}
+        : GlobalTypeRewriter(wasm), parent(parent),
+          placeholderIt(parent.descriptorsOfPlaceholders.begin()) {}
+
+      std::vector<HeapType> getSortedTypes(PredecessorGraph preds) override {
+        auto types = GlobalTypeRewriter::getSortedTypes(std::move(preds));
+        // Prefix the types with placeholders to be overwritten with the
+        // placeholder describees.
+        HeapType placeholder = Struct{};
+        types.insert(
+          types.begin(), parent.descriptorsOfPlaceholders.size(), placeholder);
+        return types;
```

Comment:
I was thinking we need a new hook for this, but I guess using `getSortedTypes` works... it has to run before the rebuilding stage.